cmd/data-scanner.go

```diff
@@ -1000,22 +1001,34 @@ func (i *scannerItem) applyMaxNoncurrentVersionLimit(ctx context.Context, o Obje
 					console.Debugf(applyVersionActionsLogPrefix+" lifecycle: %s is locked, not deleting\n", obj.Name)
 				}
 			}
+			// add this version back to remaining versions for
+			// subsequent lifecycle policy applications
+			fivs = append(fivs, fi)
 			continue
 		}
+
+		// NoncurrentDays not passed yet.
+		if time.Now().UTC().Before(obj.SuccessorModTime.AddDate(0, 0, days)) {
```

Comment:
Better to use ExpectedExpiryTime(), to respect the midnight thing

Reply:
Not sure why this is useful, are you saying that before we do not honor `NoncurrentDays` ?

Reply:
I shall see if we can use `ExpectedExpiryTme` to avoid duplication. When zero `NoncurrentDays` is specified we should expire the excess versions immediately. 

> Not sure why this is useful, are you saying that before we do not honor `NoncurrentDays` ?

This function applies only to those objects that are filtered by a rule containing `NewerConcurrentVersions`. We apply `NoncurrentDays` as before, nothing has changed.

Reply:
@vadmeste hope this addresses your comment about `ExpectedExpiryTime` - https://github.com/minio/minio/pull/13815/commits/6b80d66f68b9b20d6366d05cbb908f2b8628b25a